### PR TITLE
apply session

### PR DESCRIPTION
**Pull request by excellencetechnologies**

### CODE REVIEW

File: routes/index.js [hunk 1]
```diff
@@ -26,18 +29,37 @@ rtm.on(CLIENT_EVENTS.RTM.RTM_CONNECTION_OPENED, function () {
...
+rtm.on(RTM_CLIENT_EVENTS.RTM_CONNECTION_OPENED, function () {
+// This will send the message 'this is a test message' to the channel identified by id 'C0CHZA86Q'
+    rtm.sendMessage('text : welcome,\n username: shekhar, \n your app is running now', config.shekhar_channelId, function messageSent() {
+    });
+});
+
+
+var RTM_EVENTS = require('@slack/client').RTM_EVENTS;
+// Responds to a message with a 'hello' DM
+var to = '';
```
Comment (reactexcel) [inline routes/index.js]: these need to be removed

File: routes/index.js [hunk 2]
```diff
@@ -48,73 +70,113 @@ rtm.on(RTM_EVENTS.MESSAGE, function (message) {
...
-//////////////////////when app is started it sends a mesage to shekhar  //////////////////////////
-var RTM_CLIENT_EVENTS = require('@slack/client').CLIENT_EVENTS.RTM;
-// you need to wait for the client to fully connect before you can send messages
-rtm.on(RTM_CLIENT_EVENTS.RTM_CONNECTION_OPENED, function () {
-// This will send the message 'this is a test message' to the channel identified by id 'C0CHZA86Q'
-    rtm.sendMessage('text : welcome,\n username: shekhar, \n your app is running now', config.shekhar_channelId, function messageSent() {
-    });
+
+//     function exists(id) {
+//         var check_session = session[id] ? true : false;
```
Comment (reactexcel) [inline routes/index.js]: remove commented code. keep back in ur computer

File: routes/index.js [hunk 3]
```diff
@@ -26,18 +29,37 @@ rtm.on(CLIENT_EVENTS.RTM.RTM_CONNECTION_OPENED, function () {
...
+    });
+});
+
+
+var RTM_EVENTS = require('@slack/client').RTM_EVENTS;
+// Responds to a message with a 'hello' DM
+var to = '';
+var from = '';
+var reason = '';
+var session = {};
```
Comment (reactexcel) [inline routes/index.js]: all variables to, from, reason will be removed

session ={} will move to session.js

File: routes/index.js (hunk 2 above)
Comment (excellencetechnologies) [inline routes/index.js]: Fixed

File: routes/index.js (hunk 1 above)
Comment (excellencetechnologies) [inline routes/index.js]: Fixed

File: routes/index.js (hunk 3 above)
Comment (excellencetechnologies) [inline routes/index.js]: Fixed

File: routes/index.js [hunk 4]
```diff
@@ -48,73 +53,14 @@ rtm.on(RTM_EVENTS.MESSAGE, function (message) {
...
-//            } else {
-//                rtm.sendMessage('Invalid Days. So please try again', dm.id);
-//            }
-//        }
-//   
-    } else if (date == false && to != '' && from != '') {
-        rtm.sendMessage('Oops! Unable to understand. Please provide date in proper format (DD-MM-YYYY)', dm.id);
+    } else if (message.text == 'apply' || date == true || date == false) {
+        var id = message.user;
+        apply_.apply(message, dm, id, date, time, rtm, user);
```
Comment (reactexcel) [inline routes/index.js]: change this to leave._apply

File: service/leave/apply.js [hunk 5]
```diff
@@ -1,62 +1,65 @@
 var leave_ = require('../leaveApply');
+var session_ = require('../session');
```
Comment (reactexcel) [inline service/leave/apply.js]: change this to _session

File: service/session.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+var session = {};
+
+exports.mySession = function (cb) {
```
Comment (reactexcel) [inline service/session.js]: rename to exists

File: service/session.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        clearTimeout(session[id].timeout);
+        session[id].timeout = setTimeout(function () {
+            destory(id); //auto expire after 5sec
+        }, 50000000);
+    } else {
+//doesnt exist throw error
+    }
+};
+
+exports.set = function (session, id, key, value) {
```
Comment (reactexcel) [inline service/session.js]: remove session

File: service/leave/apply.js [hunk 8]
```diff
@@ -1,62 +1,65 @@
...
-                } else {
-                    rtm.sendMessage('Oops! Some problem occurred. We are looking into it. In the mean time you can use HR system to apply your leave', dm.id);
+exports.apply = function (message, dm, id, date, time, rtm, user, callback) {
+    exists(id);
+    function exists(id) {
+        session_.mySession(function (res) {
+            var check_session = res[id] ? true : false;
+            if (check_session == false) {
+                session_.start(res, id, time, callback);
+                rtm.sendMessage(user.name + '!' + ' can you please provide me the details \n from (DD-MM-YYYY) ', dm.id);
```
Comment (reactexcel) [inline service/leave/apply.js]: session.set(id,'command','from')

File: service/session.js (hunk 6 above)
Comment (excellencetechnologies) [inline service/session.js]: Fixed

File: routes/index.js (hunk 4 above)
Comment (excellencetechnologies) [inline routes/index.js]: Fixed

File: service/leave/apply.js (hunk 5 above)
Comment (excellencetechnologies) [inline service/leave/apply.js]: Fixed

File: service/session.js (hunk 7 above)
Comment (excellencetechnologies) [inline service/session.js]: Fixed

File: service/leave/apply.js (hunk 8 above)
Comment (excellencetechnologies) [inline service/leave/apply.js]: Fixed.

**Pull request merged**